src/index/mod.rs

```diff
@@ -0,0 +1,180 @@
+//
+// Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+
+//! The Index API provides a way to efficiently search data in the stream.
+//!
+//! The index API writes a fixed sized [`IndexRecord`] to the stream. Each [`IndexRecord`] contains the user data
+//! and a number of user defined `Fields`, the value of each `Field` has to be u64 type.
+//! An example of `Fields` is showed as below:
+//! ```no_run
+//! use pravega_client_macros::Fields;
+//!
+//! // Use Fields procedural marco to auto implement the necessary trait for MyFields.
+//! #[derive(Fields, Debug, PartialOrd, PartialEq)]
+//! struct MyFields {
+//!     time: u64, // A field
+//!     id: u64,
+//! }
+//!
+//! ```
+//!
+//! To ensure the searching efficiency in an index stream, we impose some constraints to the `Fields`:
+//! * The value of the `Field` must be monotonically increasing.
+//! * Deletion of the `Field` is not permitted.
```

Comment:
This should be put into a section about versioning and upgrades. Here it sounds like the structure itself is dynamic.